lib/discourse_algolia.rb

```diff
@@ -0,0 +1,246 @@
+      if should_index_user?(user)
+        users << user_to_object(user)
+        ids.delete(user.id)
+      end
+    end
+
+    users_index.save_objects(users) if users.present?
+    users_index.delete_objects(ids) if ids.present?
+  end
+
+  def self.process_tags!(ids)
+    return if ids.blank?
+
+    tags = []
+
+    Tag.where(id: ids).find_each do |tag|
+      if should_index_tag?(tag)
+        tags << tag_to_object(tag)
+        ids.delete(tag.id)
+      end
+    end
+
+    tags_index.save_objects(tags)  if tags.present?
+    tags_index.delete_objects(ids) if ids.present?
+  end
+
+  def self.should_index_post?(post)
+    algolia_guardian.can_see?(post)
+  end
+
+  def self.should_index_user?(user)
+    algolia_guardian.can_see?(user)
+  end
+
+  def self.should_index_tag?(tag)
+    algolia_guardian.can_see?(tag) && tag.topic_count > 0
+  end
+
+  def self.post_to_object(post)
+    {
+      objectID: post.id,
+      url: post.url,
+      post_id: post.id,
+      post_number: post.post_number,
+      created_at: post.created_at.to_i,
+      updated_at: post.updated_at.to_i,
+      reads: post.reads,
+      like_count: post.like_count,
+      image_url: post.image_url,
+      word_count: post.word_count,
+      content: Nokogiri::HTML5.fragment(post.cooked).text,
+      user: {
+        id: post.user.id,
+        url: "/u/#{post.user.username_lower}",
+        name: post.user.name,
+        username: post.user.username,
+        avatar_template: post.user.avatar_template,
+      },
+      topic: {
+        id: post.topic.id,
+        url: post.topic.url,
+        title: post.topic.title,
+        views: post.topic.views,
+        slug: post.topic.slug,
+        like_count: post.topic.like_count,
+        tags: post.topic.tags.map(&:name),
+      },
+      category: {
+        id: post.topic.category.id,
+        url: post.topic.category.url,
+        name: post.topic.category.name,
+        color: post.topic.category.color,
+        slug: post.topic.category.slug,
+      }
+    }
+  end
+
+  def self.user_to_object(user)
+    {
+      objectID: user.id,
+      url: "/u/#{user.username_lower}",
+      name: user.name,
+      username: user.username,
+      avatar_template: user.avatar_template,
+      bio_raw: user.user_profile.bio_raw,
+      post_count: user.post_count,
+      badge_count: user.badge_count,
+      likes_given: user.user_stat.likes_given,
+      likes_received: user.user_stat.likes_received,
+      days_visited: user.user_stat.days_visited,
+      topic_count: user.user_stat.topic_count,
+      posts_read: user.user_stat.posts_read_count,
+      time_read: user.user_stat.time_read,
+      created_at: user.created_at.to_i,
+      updated_at: user.updated_at.to_i,
+      last_seen_at: user.last_seen_at,
+    }
+  end
+
+  def self.tag_to_object(tag)
+    {
+      objectID: tag.id,
+      url: tag.url,
+      name: tag.name,
+      topic_count: tag.topic_count,
+    }
+  end
+
+  def self.posts_index
+    @posts_index ||= begin
```

Comment:
I think it would be wort extracting this logic:

```
def build_index(name, setting)
  index = algolia_index(name)
  index.set_settings(setting) unless index.exists?
  index
end

def self.users_index
  @users_index ||= build_index('discourse-users', USER_SETTINGS)
end
```